sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs

```diff
@@ -225,6 +226,33 @@ public class ServiceBusReceiver : IAsyncDisposable
             return messages;
         }
 
+        /// <summary>
+        /// Receives messages as an asynchronous enumerable from the entity using <see cref="ReceiveMode"/> mode.
+        /// <see cref="ReceiveMode"/> defaults to PeekLock mode. Messages will be received from the entity as
+        /// the IAsyncEnumerable is iterated. If no messages are available, this method will continue polling
+        /// until messages are available, i.e. it will never return null.
+        /// </summary>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the
+        /// request to cancel the operation.</param>
+        /// <returns>The message received.</returns>
+        public virtual async IAsyncEnumerable<ServiceBusReceivedMessage> ReceiveMessagesAsync(
```

Comment:
Should we consider adding an overload that accepts `maxWaitTime` in order to allow for returning control to the caller when no message is available after a period of time?   This is often helpful for heartbeat scenarios.

In its current form, the only way a caller would be able to exit the loop would be to set the cancellation token to expire after an absolute period, whether or not messages are available or to signal the cancellation token from another thread.

Reply:
I think that is an interesting idea - we would return null if there was no message available in the time period, if I'm understanding correctly. The thing is we could technically still do that here. What I am getting at is that we may need more than just a separate overload for behavior that is this fundamentally different. I avoided including the MaxWaitTime with the current behavior as I could see it causing significant confusion, and the user is unlikely to need this level of tuning when we are still just waiting for the next message. I would vote to wait for customer feedback before adding this feature.

Reply:
>  we would return null if there was no message available in the time period, if I'm understanding correctly. 

Yes, that's exactly what I'm proposing.  We use this pattern in Event Hubs based on feedback from UX studies;  it allows us to maintain the same `maxWaitTime` in the iterator as the `Receive` method has and kept developers from having to break out of a loop and then restart just because they wanted to log every XX minutes but no messages were in the system.

The "emit whether there is an event or not" didn't go over well in studies.  It differed from the `Receive` behavior in that the loop was constant and those that preferred to just wait forever for an event were annoyed at having to filter the nulls.

Reply:
Interesting.. @KieranBrantnerMagee @conniey @richardpark-msft do your generators/iterators expose the ability to return a null message after the specified maxWaitTime in addition to the ability to just wait indefinitely for the next message?

Reply:
We don't in JS but I did leave this comment a bit ago to myself to talk about it at some point:

```typescript
    // In EventHubs we've had a concept of "punctuation" (thanks @jsquire) that
    // allows the user, when working in a model like this, to get a periodic "no message
    // arrived in this window of time" notification.
    //
    // TODO: do we want this same behavior for ServiceBus?
    if (messages.length === 0) {
      continue;
    }
```

Reply:
I guess my question is, how is this different from just using the regular receive and looping yourself. I thought the whole benefit of the async enumerable is that you don't have to handle cases where you don't get any messages. If customers want this heartbeat, they can just do:
```c# 
while (true)
{
  var msg = await receiver.ReceiveMessage(maxWaitTime: TimeSpan.FromSeconds(10));
  // logic involving msg (can be null)
}

I think Event Hubs land, they might do some more complex stuff involving publishing events to channels to try to ensure a consistent stream of messages, so perhaps that is why it made sense there. In Service Bus, we can't do this because we need to worry about the message lock.

Reply:
It's a fair point;  the main difference is that there is no `ReceiveAsync` in Event Hubs.   The iterator is the only way to read events using the standard consumer client.   I hadn't taken that into account.   Your alternate option fills the case that we had a gap for.   

Given the new context, I withdraw my suggestion.   I think it's cleaner the way that you've got it.

Reply:
I wouldn't call it a `waitTime`. Something more accurate like `idleTime` would be great because it explains what you're waiting for. Or, even `idleTimeout` would be great, because it says exactly what it is. It times out after being idle for X amount of time.

Reply:
It's a good thought, but there's quite a bit of precedence in the Azure Messaging space for `maximum wait time`, going back to the track zero library for Service Bus/Event Hubs and being carried forward through track one for both and then into Event Hubs track two.   My concern with renaming it here would be diverging from both history and the same concept in another track two library.